### PR TITLE
Add document about single region 1PC

### PR DESCRIPTION
**Pull request by MyonKeminta**
Signed-off-by: MyonKeminta <MyonKeminta@users.noreply.github.com>

This PR adds a draft document for single region 1PC. This PR  is also  expected to be the place to do related discussions.

### CODE REVIEW

**Comment by MyonKeminta** [comment]
@cfzjywxk @sticnarf @youjiali1995 @coocood PTAL. How do you think the future plan should be?

**Comment by sticnarf** [comment]
The performance improvement is promising. I think we can get started with dealing with the corner cases and compatibility issues.

**Comment by sticnarf** [comment]
And I think we can commit your 1PC implementation to the codebase now as long as it's not enabled.

File: design/async-commit/single-region-1pc.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+Possible solution: When trying committing a transaction with 1PC, find a ts `one_pc_max_commit_ts` before which we can guarantee that the schema version can't change, and send it to TiKV. TiKV will reject committing if the calculated `commit_ts` exceeds the `one_pc_max_commit_ts`.
+
+### CDC compatibility problem
+
+CDC syncs data from TiKV by observing applying events. Prewrites and commits are distinguished, and CDC will use these events to compose complete transactions and then send them to the downstream. So when 1PC is enabled, there need to be some way for CDC to distinguish if a apply event is caused by 1PC committing, otherwise CDC will expect that a commit event must has a corresponding prewrite event to compose a complete transaction.
+
+Possible solutions:
```
Comment (coocood) [inline design/async-commit/single-region-1pc.md:37]: Firstly, we should have a way that disables 1PC when CDC is connected.

Another solution is to construct a Prewrite event before Commit event for CDC.

Comment (sticnarf) [reply design/async-commit/single-region-1pc.md:37]: Will 1PC be shipped with the problem unsolved? Otherwise I think we can allow enabling 1PC only after resolving these problems, then we won't need to automatically disable 1PC when CDC is connected.

**Comment by youjiali1995** [comment]
Land it in v5.0. There isn't any unsolvable problem. We can polish and test it with async-commit in the next few months.

File: design/async-commit/single-region-1pc.md (hunk 1 above)
Comment (coocood) [reply design/async-commit/single-region-1pc.md:37]: So we can enable 1PC by default on the master branch.
It makes POC much easier.

Comment (MyonKeminta) [reply design/async-commit/single-region-1pc.md:37]: I'm not sure if enabling it by default is a good idea unless we've done enough tests and it's stable enough.

Comment (coocood) [reply design/async-commit/single-region-1pc.md:37]: enable it by default can make the feature mature more quickly.

**Comment by MyonKeminta** [comment]
It's exciting that you agree landing it in 5.0. But I'm still afraid that there might still be other corner cases or compatibility issues I didn't notice.

**Comment by cfzjywxk** [comment]
> Land it in v5.0. There isn't any unsolvable problem. We can polish and test it with async-commit in the next few months.

Agree to land it in v5.0, we could limit the usage scenarios for it and do enough tests. We should take care of compatible problems and do not create extra work for tools team.

**Comment by MyonKeminta** [comment]
@youjiali1995 Do you know if this affects TiFlash?

**Comment by youjiali1995** [comment]
TiFlash may have some protective actions like CDC doing, but it's easy to solve. Besides it, I think no compatibility issues with TiFlash.

File: design/async-commit/single-region-1pc.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
+# Single Region 1PC
+
+For transactions that affect only one region, or more strictly speaking, that can be prewritten with only one prewrite request, can be committed directly while the prewrite request is being handled, so the commit phase can be totally removed. Therefore we can get less latency and more throughput. For TiDB, indices and rows are unsally not in a same region, **so this optimization only works under a limited amount of scenarios, like sysbench oltp_update_non_index**. But in a suitable scenario, it gains significantly better performance.
+
+In the old days we did some attempt to implement single region 1PC for single-region transactions, but due to incompatibility and some technical problem it was delayed. [There is some document left at that time (note that it's very outdated)](https://docs.google.com/document/d/1Vkk8LpYbXaQ0ualdFFH35V6mv9c9RsWJu2s6nhJz9E4/edit). But now since async commit, which meets and solved mostly the same problem as 1PC, is implemented, we can continue supporting 1PC with much less effort to make.
```
Comment (cfzjywxk) [inline design/async-commit/single-region-1pc.md:5]: Maybe we could extend it to 1PC for single-store transactions with some placement usages in the future.

Comment (zhangjinpeng1987) [reply design/async-commit/single-region-1pc.md:5]: Please remove the google document.

File: design/async-commit/single-region-1pc.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# Single Region 1PC
+
+For transactions that affect only one region, or more strictly speaking, that can be prewritten with only one prewrite request, can be committed directly while the prewrite request is being handled, so the commit phase can be totally removed. Therefore we can get less latency and more throughput. For TiDB, indices and rows are unsally not in a same region, **so this optimization only works under a limited amount of scenarios, like sysbench oltp_update_non_index**. But in a suitable scenario, it gains significantly better performance.
+
+We used to think about supporting 1PC before but stopped after meeting many difficulties. Now since async commit, which meets and solved mostly the same problem as 1PC, is implemented, we can continue supporting 1PC with much less effort to make.
```
Comment (MyonKeminta) [inline design/async-commit/single-region-1pc.md:5]: From @cfzjywxk :
> Maybe we could extend it to 1PC for single-store transactions with some placement usages in the future.

File: design/async-commit/single-region-1pc.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  * Memory lock
+  * Replica read problem
+* Problems that we do not care anymore:
+  * Binlog incompatibility
+
+## Basic Design
+
+Since async commit is implemented, it's not difficult to implement a working 1PC. However, it's still hard to make it perfectly correct and compatible with other components.
+
+* When committing a transaction, if TiDB finds that the prewrite phase can be done with only one single request, the transaction is allowed to be committed with 1PC protocol. A field named `try_one_pc` in the prewrite request will be set to let TiKV know that 1PC is available for this transaction.
```
Comment (nrc) [inline design/async-commit/single-region-1pc.md:19]: When would this be set or not set? If and only if the transaction only touches one region/store?

File: design/async-commit/single-region-1pc.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+Possible solution: When trying committing a transaction with 1PC, find a ts `one_pc_max_commit_ts` before which we can guarantee that the schema version can't change, and send it to TiKV. TiKV will reject committing if the calculated `commit_ts` exceeds the `one_pc_max_commit_ts`.
+
+### CDC compatibility problem
+
+CDC syncs data from TiKV by observing applying events. Prewrites and commits are distinguished, and CDC will use these events to compose complete transactions and then send them to the downstream. So when 1PC is enabled, there need to be some way for CDC to distinguish if a apply event is caused by 1PC committing, otherwise CDC will expect that a commit event must has a corresponding prewrite event to compose a complete transaction.
+
+Possible solutions:
+1. Passing a `is_1pc` flag from txn layer to apply, just like how `TxnExtra` (which is used to support CDC outputting old value) was written previously. It would be ugly.
```
Comment (nrc) [inline design/async-commit/single-region-1pc.md:36]: I think this is the least ugly solution :-) 2 is very fragile, 3 seems wasteful and make write records more complex.

**Review by nrc — COMMENTED** [review]
The proposal looks good. I'd be wary of shipping in 5.0 - async commit is already a really large, high risk change and I'd minimise other changes to the transaction system. Furthermore, although the perf improvements for 1pc are good, they are really good compared to 2pc and ok compared to async commit.

File: design/async-commit/single-region-1pc.md (hunk 4 above)
Comment (MyonKeminta) [reply design/async-commit/single-region-1pc.md:19]: It will be set iff the prewrite can be done with only one request. Typically it means the transaction affects only one region. But if there's too many affected keys in a region, it may need to be prewritten with more than one requests, in which case 1PC is not applicable.

File: design/async-commit/single-region-1pc.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
+# Single Region 1PC
+
+For transactions that affect only one region, or more strictly speaking, that can be prewritten with only one prewrite request, can be committed directly while the prewrite request is being handled, so the commit phase can be totally removed. Therefore we can get less latency and more throughput. For TiDB, indices and rows are unsally not in a same region, **so this optimization only works under a limited amount of scenarios, like sysbench oltp_update_non_index**. But in a suitable scenario, it gains significantly better performance.
```
Comment (ekexium) [inline design/async-commit/single-region-1pc.md:3]:
```suggestion
For transactions that affect only one region, or more strictly speaking, that can be prewritten with only one prewrite request, can be committed directly while the prewrite request is being handled, so the commit phase can be totally removed. Therefore we can get less latency and more throughput. For TiDB, indices and rows are usually not in a same region, **so this optimization only works under a limited amount of scenarios, like sysbench oltp_update_non_index**. But in a suitable scenario, it gains significantly better performance.

```

**Comment by andylokandy** [comment]
@MyonKeminta Do we have any updates?
